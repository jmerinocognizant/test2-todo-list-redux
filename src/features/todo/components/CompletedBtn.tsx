interface props {
    completed: boolean,
    onClickHandler: () => void,
}

export const CompletedBtn = ({completed,onClickHandler}:props) => {
  return (
    <>
        {
        completed?
            <button aria-label="btn-completed" className="btn btn-block btn-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-check" ></i>
            </button>
            :
            <button aria-label="btn-completed" className="btn btn-block btn-outline-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-check" ></i>
            </button>
        }    
    </>
  )
}
