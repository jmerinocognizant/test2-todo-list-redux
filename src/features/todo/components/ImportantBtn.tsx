interface props {
    important: boolean,
    onClickHandler: () => void,
}

export const ImportantBtn = ({important,onClickHandler}:props) => {
  return (
    <>
        {
        important?
            <button aria-label="btn-important" className="btn btn-block btn-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-exclamation" ></i>
            </button>
            :
            <button aria-label="btn-important" className="btn btn-block btn-outline-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-exclamation" ></i>
            </button>
        }    
    </>
  )
}
