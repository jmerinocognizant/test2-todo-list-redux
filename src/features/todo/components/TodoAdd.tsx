import { useForm } from "../hooks/useForm";
import { Todo } from "../interfaces/interfaces";
import { useAppDispatch } from '../../../app/hooks';
import { addTodo } from '../store/todoSlice';

export const TodoAdd = () => {
  
  const dispatch = useAppDispatch();

  const { description, onInputChange, onResetForm } = useForm({
    description: ''
    });

  const onFormSubmit = ( event : React.FormEvent<HTMLFormElement> ) => {
    event.preventDefault();
    if ( description.length <= 1 ) return;

    const newTodo:Todo = {
        id: new Date().getTime(),
        description: description,
        completed: false,
        important: false,
    }

    dispatch(addTodo(newTodo));
    onResetForm();
  }
  
  return (
    <>
      <div className="container-fluid">
          <form aria-label="form" onSubmit = {onFormSubmit }>

          <div className="row">
            <div className="col">
              <input 
                      type="text" 
                      placeholder="Add something to do..."
                      className="form-control align-top"
                      name="description"
                      value={ description }
                      onChange={ onInputChange }
                  />
              </div>
            <div className="col-auto mt-0">      
              <button
                  type="submit" 
                  className="btn btn btn-primary align-top"
              >
                  Save
              </button>
            </div>
            </div>  
          </form>
      </div>
    </>
  )
}
