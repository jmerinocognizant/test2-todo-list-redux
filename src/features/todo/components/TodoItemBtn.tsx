interface props {
    checked: boolean,
    classNameIcon: string,
    onClickHandler: () => void,
}

export const TodoItemBtn = ({checked,classNameIcon,onClickHandler}:props) => {
    const btnClassName = checked? "btn btn-block btn-primary item-btn":"btn btn-block btn-outline-primary item-btn"
    
  return (
    <>
         <button aria-label="btn-todo" className={btnClassName} onClick={onClickHandler}>
                <i className={classNameIcon} ></i>
         </button>
    </>
  )
}
