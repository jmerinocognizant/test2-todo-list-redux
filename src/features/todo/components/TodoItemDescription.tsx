
interface props {
    description: string,
    onDoubleClick: () => void,
}

export const TodoItemDescription = ({description,onDoubleClick}:props) => {
  return (
    <>
        <div aria-label="item-description" className="w-100 p-0" onDoubleClick={onDoubleClick}>
            <span  className="h6 v-top ">{description}</span>
        </div>
    </>
  )
}
