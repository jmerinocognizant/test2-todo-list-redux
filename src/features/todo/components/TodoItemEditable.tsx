import { useState } from "react";

interface props {
    description: string,
    onSaveClick: (text:string) => void,
}
export const TodoItemEditable = ({description,onSaveClick}:props) => {

  const [textDescription, setTextDescription] = useState(description);
  const [disabled, setDisabled] = useState(false);

  const onSaveClickHandler = () => {
    if ( description.length <= 1 ) return;
    onSaveClick(textDescription);
  }

  const onInputChange = (text:string) => {
    setTextDescription(text);
    setDisabled(text.length<=1);
  }

  return (
    <>
        <div className="container-fluid p-0 ">
            <div className="row" >
                <div className="col">
                    <input 
                        type="text" 
                        className="form-control align-top"
                        name="description"
                        value={ textDescription }
                        onChange={ (event) => onInputChange(event.target.value) }
                    />
                </div>
                <div className="col-auto mt-0">
                    <button 
                        aria-label="btn-save"
                        className="btn btn-primary align-top"
                        disabled={disabled}
                        onClick={onSaveClickHandler}>
                        save
                    </button>
                </div>
            </div>
        </div>
    </>
  )
}
