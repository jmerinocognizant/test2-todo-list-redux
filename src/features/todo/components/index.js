export * from './CompletedBtn';
export * from './ImportantBtn';
export * from './TodoAdd';
export * from './TodoItem';
export * from './TodoItemDescription';
export * from './TodoItemEditable';
export * from './TodoList';