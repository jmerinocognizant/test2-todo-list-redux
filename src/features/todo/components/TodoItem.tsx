import { Todo } from "../interfaces/interfaces"
import { CompletedBtn } from "./CompletedBtn";
import { ImportantBtn } from "./ImportantBtn";
import { TodoItemDescription } from "./TodoItemDescription";
import { TodoItemEditable } from "./TodoItemEditable";
import { useAppDispatch } from '../../../app/hooks';
import {
    removeTodo,
    updateTodo,
    setEditable,
  } from '../store/todoSlice';
import { TodoItemBtn } from "./TodoItemBtn";
  
interface props {
    todo: Todo,
    editable: boolean,
}

export const TodoItem = ( {todo,editable}:props) => {
  
    const dispatch = useAppDispatch();

    const onDoubleClick = () => {
        dispatch(setEditable(todo.id));
    }

    const onRemoveTodo = () => {
        dispatch(removeTodo(todo.id));
    }

    const handleClickImportant = () => {
        const todoUpdated = {...todo,
                            important: !todo.important };
        dispatch(updateTodo(todoUpdated));
    }

    const handleClickCompleted = () => {
        const todoUpdated = {...todo,
            completed: !todo.completed };
        dispatch(updateTodo(todoUpdated));
    }

    const onSaveClickHandler = (textDescription:string) =>{
        const todoUpdated = {...todo,
            description: textDescription };
        dispatch(updateTodo(todoUpdated));
        dispatch(setEditable(0));
    }

    return (
        <>    
            < hr/>
            <div aria-label="todo-item" className="container-fluid todo-item">
                <div className="row">
                    <div className="col-auto mt-0  v-top">
                        {/* <CompletedBtn completed={todo.completed} onClickHandler={handleClickCompleted}/>  */}
                        <TodoItemBtn aria-label="btn-completed" checked={todo.completed} classNameIcon="fa-solid fa-check" onClickHandler={handleClickCompleted} />
                    </div>
                    <div className="col-auto mt-0  v-top">
                        {/* <ImportantBtn important={todo.important} onClickHandler={handleClickImportant}/> */}
                        <TodoItemBtn aria-label="btn-important" checked={todo.important} classNameIcon="fa-solid fa-exclamation" onClickHandler={handleClickImportant} />
                    </div>
                    
                    <div className="col">
                        {
                        editable ?(
                            <TodoItemEditable description = {todo.description} onSaveClick={onSaveClickHandler}/>  
                        ):(  
                            <TodoItemDescription description={todo.description} onDoubleClick={onDoubleClick}/>      
                        )
                        }
                    </div>
                    <div className="col-auto mt-0">
                        <button aria-label="btn-remove" className="btn btn-danger align-top" onClick={onRemoveTodo}>
                            remove
                        </button>
                    </div>
                </div>
            </div>
        </>
    )
}
