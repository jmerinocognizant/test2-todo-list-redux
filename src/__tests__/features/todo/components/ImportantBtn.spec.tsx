import { fireEvent, render, screen } from '@testing-library/react';
import { ImportantBtn } from '../../../../features/todo/components';

describe('Pruebas en <ImportantBtn />', () => { 

    const onClickHandler = jest.fn();

    test('debe de hacer match con el snapshot', () => {
        const { container } = render(
             < ImportantBtn important={false} onClickHandler={onClickHandler}/>
        );

        expect( container ).toMatchSnapshot();
    });

    test('debe de mostrar un botón fa-exclamation sin seleccionar', () =>{
        const { container, getByText } = render(
            < ImportantBtn important={false} onClickHandler={onClickHandler}/>
       );

       const button = container.querySelector('button');
       expect(button).toHaveClass('btn-outline-primary');
       expect(button?.firstChild).toHaveClass('fa-exclamation');
    })


    test('debe de mostrar el botón seleccionado', () =>{
        const { container, getByText } = render(
            < ImportantBtn important={true} onClickHandler={onClickHandler}/>
       );

       const button = container.querySelector('button');
       expect(button).toHaveClass('btn-primary');
    })

    test('debe de realizar realizar la llamada a la función onClick', () =>{
        const { container, getByText } = render(
            < ImportantBtn important={true} onClickHandler={onClickHandler}/>
       );

       fireEvent.click(screen.getByRole('button',{name: 'btn-important'}));

       expect( onClickHandler ).toHaveBeenCalled();
       

    })

 });

 