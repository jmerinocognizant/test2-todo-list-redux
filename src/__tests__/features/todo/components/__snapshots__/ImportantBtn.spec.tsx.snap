// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ImportantBtn /> debe de hacer match con el snapshot 1`] = `
<div>
  <button
    aria-label="btn-important"
    class="btn btn-block btn-outline-primary item-btn"
  >
    <i
      class="fa-solid fa-exclamation"
    />
  </button>
</div>
`;
